curve25519-dalek/src/backend/serial/u64/field_lemmas/compute_q_lemmas.rs

```diff
@@ -0,0 +1,723 @@
+
+    assert((q4 * pow2(51) as int + r4 - q3) * pow2(204) as int == q4 * pow2(51) as int * pow2(
+        204,
+    ) as int + r4 * pow2(204) as int - q3 * pow2(204) as int) by {
+        lemma_mul_is_distributive_sub_other_way(pow2(204) as int, q4 * pow2(51) as int + r4, q3);
+        lemma_mul_is_distributive_add_other_way(pow2(204) as int, q4 * pow2(51) as int, r4);
+    }
+
+    // Now LHS equals (after substituting the expansions):
+    // = q0*2^51 + r0 - 19
+    //   + q1*2^51*2^51 + r1*2^51 - q0*2^51
+    //   + q2*2^51*2^102 + r2*2^102 - q1*2^102
+    //   + q3*2^51*2^153 + r3*2^153 - q2*2^153
+    //   + q4*2^51*2^204 + r4*2^204 - q3*2^204
+    //   + 19
+
+    // Use the power relationships to simplify products
+    assert(q1 * pow2(51) as int * pow2(51) as int == q1 * pow2(102) as int) by {
+        lemma_mul_is_associative(q1, pow2(51) as int, pow2(51) as int);
+    }
+    assert(q2 * pow2(51) as int * pow2(102) as int == q2 * pow2(153) as int) by {
+        lemma_mul_is_associative(q2, pow2(51) as int, pow2(102) as int);
+    }
+    assert(q3 * pow2(51) as int * pow2(153) as int == q3 * pow2(204) as int) by {
+        lemma_mul_is_associative(q3, pow2(51) as int, pow2(153) as int);
+    }
+
+    // Now we can see the cancellations more clearly:
+    // = (q0*2^51 - q0*2^51) + (q1*2^102 - q1*2^102) + (q2*2^153 - q2*2^153) + (q3*2^204 - q3*2^204)
+    //   + q4*2^51*2^204 + r0 + r1*2^51 + r2*2^102 + r3*2^153 + r4*2^204 + (-19 + 19)
+    // = q4*2^51*2^204 + r0 + r1*2^51 + r2*2^102 + r3*2^153 + r4*2^204
+
+    // The SMT solver should now see this is pure linear arithmetic
+}
+
+/// Direct proof of telescoping division for radix-51 representation
+/// Uses repeated substitution to show q4 = (as_nat(limbs) + 19) / 2^255
+///
+/// Proof strategy:
+/// 1. Express as_nat(limbs) + 19 as a sum using radix-51: Σ limbs[i] * 2^(51*i) + 19
+/// 2. Substitute each limb using the division theorem equations (from requires clause)
+/// 3. Expand and observe that intermediate q_i terms telescope (cancel out):
+///    - q0 appears as: +q0*2^51 - q0*2^51 = 0
+///    - q1 appears as: +q1*2^102 - q1*2^102 = 0  (and so on)
+/// 4. After cancellation: value = q4 * 2^255 + remainder, where remainder < 2^255
+/// 5. By uniqueness of division, q4 = value / 2^255
+pub proof fn lemma_radix51_telescoping_direct(
+    limbs: [u64; 5],
+    q0: int,
+    q1: int,
+    q2: int,
+    q3: int,
+    q4: int,
+    r0: int,
+    r1: int,
+    r2: int,
+    r3: int,
+    r4: int,
+)
+    requires
+        limbs[0] as int + 19 == q0 * pow2(51) as int + r0,
+        limbs[1] as int + q0 == q1 * pow2(51) as int + r1,
+        limbs[2] as int + q1 == q2 * pow2(51) as int + r2,
+        limbs[3] as int + q2 == q3 * pow2(51) as int + r3,
+        limbs[4] as int + q3 == q4 * pow2(51) as int + r4,
+        0 <= r0 < pow2(51) as int,
+        0 <= r1 < pow2(51) as int,
+        0 <= r2 < pow2(51) as int,
+        0 <= r3 < pow2(51) as int,
+        0 <= r4 < pow2(51) as int,
```

Comment:
I'm generally not a huge fan of this style of lemma parameters, instead of `q`s and `r`s as arguments, I'd rather see them computed, since it's clearly the case that 
`q_i = (limbs[i] + q_{i-1}) / 2^51` and `r_i = (limbs[i] + q_{i-1}) % 2^51` (`q_{-1} := 19`)
The only real argument here is `limbs`.

OTOH, formulating `q4` is quite verbose, so the ensures would be quite unreadable. Just making a note to possibly kick off a style discussion for the style guide.